### PR TITLE
persist: add SQL to expose shard internals for debugging and QA

### PR DESCRIPTION
**Pull request by danhhz**
This adds an mz_system only `INSPECT SHARD <id>` command for debugging and QA.

Example:

```
materialize=> INSPECT SHARD 'u1';
state
------------------
{"batches": ... }
(1 row)
```

As discussed in https://github.com/MaterializeInc/materialize/issues/19102#issuecomment-1540789978, it's possible but tricky to make this work in a way that lets us use normal SQL jsonb operators on it. This first version chooses to do the strait forward implementation, which pushes any processing of the json into the client. If this command proves to be sufficiently useful and this limitation proves to be a sufficiently annoying papercut, then we can revisit.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
@mjibson after shaving _many_ yaks, I have returned with the method we discussed in https://github.com/MaterializeInc/materialize/issues/19102#issuecomment-1540789978 to dump a persist shard's state as JSON given the collection id (we said shard id but collection id was easy and is better I think). if you're still up for hooking this up in SQL as some sort of new `INSPECT` command, that'd be lovely. if not, no biggie, I can probably muddle through myself with some pointers

**Comment by mjibson** [comment]
`INSPECT` commit here: https://github.com/mjibson/materialize/commit/4134cd32370befb9f526014a9ec682e0b27cc92e

Didn't want to push to this branch in case you were mid pushing fixes.

**Comment by danhhz** [comment]
amazing! \o/ feel free to push to this branch, I made it for exactly that purpose

File: src/adapter/src/coord/sequencer.rs [hunk 1]
```diff
@@ -200,6 +200,10 @@ impl Coordinator {
                 let result = self.sequence_show_variable(ctx.session(), plan);
                 ctx.retire(result);
             }
+            Plan::InspectShard(plan) => {
+                let (tx, _, session, _) = ctx.into_parts();
+                tx.send(self.sequence_inspect_shard(&session, plan).await, session);
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:205]: This is going to need a rebase on main to do  `ctx.retire` here.

File: src/adapter/src/coord/sequencer/inner.rs [hunk 2]
```diff
@@ -1480,6 +1482,27 @@ impl Coordinator {
...
+                rbac::UnauthorizedError::MzSystem {
+                    action: "inspect".into(),
+                },
+            ));
+        }
+        let state = self
+            .controller
+            .storage
+            .inspect_persist_state(plan.id)
+            .await?;
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer/inner.rs:1501]: Does this future have to be await'd in the same task? Can the future be moved to another task? We can probably leave this for now, but might be worth exploring that in the future to keep things off the coord thread when possible. I can do that in a followup.

File: src/adapter/src/coord/sequencer/inner.rs [hunk 3]
```diff
@@ -1480,6 +1482,27 @@ impl Coordinator {
         Ok(send_immediate_rows(vec![row]))
     }
 
+    pub(super) async fn sequence_inspect_shard(
+        &self,
+        session: &Session,
+        plan: InspectShardPlan,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        if !session.user().is_internal() {
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer/inner.rs:1490]: Not thrilled about this rbac special case here, but probably sufficient for now. I can improve later.

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/coord/sequencer.rs (hunk 1 above)
Comment (mjibson) [reply src/adapter/src/coord/sequencer.rs]: lgtm

**Review by pH14 — APPROVED** [review]
LGTM on the persist end, can't wait for this! Thank you @mjibson for the assist here too.

**Comment by danhhz** [comment]
TFTRs!

**Pull request merged**